### DISCUSSION TITLE
How do I embed markdown mermaid.js diagrams?

### DISCUSSION THREAD

**Discussion by fire** [quarto-dev/quarto-cli#5256] [Q&A]
How do I embed markdown mermaid.js diagrams? I want to embed a diagram like: https://github.com/V-Sekai/manuals/issues/460

**Comment by cderv** [comment]
You should start by the documentation part about Diagrams in Quarto: https://quarto.org/docs/authoring/diagrams.html

Did you already give it go and is something not clear enough ?

**Reply by AleksandarMedarevic** [reply]
Dear cderv, I have a similar question. I have created diagram in Quart and Mermaid. After rendering it provides really simple and clear plot. However, after I upload it on GitHub and using GitHubPages, I see only code chunk.  I trey to solve a problem. Probably it should be somethin in yamal file, bit without success. Could you please give me some hint. Aleksandar

**Reply by cderv** [reply]
Please do post a new question in Discussion with an example (or a link to a repo) and we could probably help you ! Thanks

**Comment by fire** [comment]
The `{mermaid}` tag is incompatible with `mermaid` tag on Github.

```{mermaid}
flowchart LR
  A[Hard edge] --> B(Round edge)
  B --> C{Decision}
  C --> D[Result one]
  C --> E[Result two]
```

```mermaid
flowchart LR
  A[Hard edge] --> B(Round edge)
  B --> C{Decision}
  C --> D[Result one]
  C --> E[Result two]
```

**Reply by cderv** [reply]
So I am guessing your question is : How can I produce github markdown output (using `format: gfm` probably ?) that will leverage the Github Flavor Markdown mermaid supported feature (https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams) ? 

If that is that, you can use markdown syntax directly 
````markdown
---
title: 'mermaid for github'
format: gfm
---

```mermaid
flowchart LR
  A[Hard edge] --> B(Round edge)
  B --> C{Decision}
  C --> D[Result one]
  C --> E[Result two]
```
````

Would produce this .md file 
````markdown
# mermaid for github

``` mermaid
flowchart LR
  A[Hard edge] --> B(Round edge)
  B --> C{Decision}
  C --> D[Result one]
  C --> E[Result two]
```

````

that renders on Github as you shared. 

Really, I am guessing - I am really unsure to correctly understand what you are looking for as you don't give much detail on the context. Quarto supports mermaid the way described in the docs, but I have feeling this is not what you want, but I don't know for sure from your two comments. 

Can you give more context ?

**Reply by fire** [reply]
I have a workflow where I use a github issue form and then transform it into a quatro `.md` document.

You're saying that I can use your suggestion and it'll work. Will try.

```
---
title: 'mermaid for github'
format: gfm
---

```mermaid
flowchart LR
  A[Hard edge] --> B(Round edge)
  B --> C{Decision}
  C --> D[Result one]
  C --> E[Result two]
```
```

**Reply by fire** [reply]
![image](https://user-images.githubusercontent.com/32321/233430451-35dc12fd-2a55-4586-bcd2-93fab752831c.png)

**Reply by fire** [reply]
Didn't work.

https://v-sekai.github.io/manuals/decisions/20230419-package-godot-for-pkgin.html

![image](https://user-images.githubusercontent.com/32321/233430709-c3643489-b233-4563-8d4b-e26a330d7e72.png)

**Reply by cderv** [reply]
We are round tripping in the discussion because I am trying to help but probably the context explanation of the situation is not clear enough for me to understand. That is why we asked usually for clear examples. 

It required me to spend some time to investigate following the breadcrumbs you left, so I am guessing and hopefully it will help, but I may also be still misunderstanding what you are doing. 

* I would guess that you are asking in the context of this workflow: https://github.com/V-Sekai/manuals/blob/master/.github/workflows/publish-quarto.yml

* It seems you are building a site which contains `.md` file that you export from issues. The one you shared seems to be https://github.com/V-Sekai/manuals/blob/master/decisions/20230419-package-godot-for-pkgin.md?plain=1 

* You then run [`quarto render` on the all project ](https://github.com/V-Sekai/manuals/blob/master/.github/workflows/publish-quarto.yml#L58-L59) - this means the `.md` files will be rendered to HTML by Quarto using the [`format: html` you defined ](https://github.com/V-Sekai/manuals/blob/master/_quarto.yml#L32-L37). This docs contains [a mermaid code chunk expressed using `GFM` syntax. ](https://github.com/V-Sekai/manuals/blob/de8c2db4733a590c657af1c65b3066b7ef3d90e9/decisions/20230419-package-godot-for-pkgin.md?plain=1#L26-L39)

* You then get a HTML version of your document, but you get only a [`<pre>` of class mermaid](https://github.com/V-Sekai/manuals/blob/gh-pages/docs/decisions/20230419-package-godot-for-pkgin.html#L168-L180). 

I am guessing this is the context right ? (which you could have explained). 

If so, then you get the expect results.

* You are converting from `.md` file, so no computation engine can be used, only Tex content using markdown syntax. 
* You are using [source code](https://quarto.org/docs/authoring/markdown-basics.html#source-code) syntax, which gives you source code html conversion. No mermaid processing is expected when using a source code syntax, and not an engine. 
* To get[ diagram feature from Quarto](https://quarto.org/docs/authoring/diagrams.html), you need to use `.qmd` file and not `.md` so that engine block can be used - here you want this for your block to be converted to HTML with correct resources. 
````markdown
```{mermaid}
...
```
````

This is Quarto feature, so you need to use Quarto document, and expected Quarto syntax. You need to add some processing script before rendering to convert to `.qmd` the file that will need to have the mermaid engine running for correct HTML output. 


Am I in the right context ? Does it helps understand what happens ? What are you expecting things to work ?

**Reply by fire** [reply]
You are correct.

You're saying my workflow only works when quarto is expecting a Quarto document with Quarto syntax.

I wasn't aware of this difference. So I was confused.

I'll have more time to review your instruction by tomorrow.

**Reply by fire** [reply]
I was expecting things the default syntax for `.md` to be github markdown.

**Reply by fire** [reply]
To restate: I was expecting the default syntax for .md to be GitHub markdown in quatro. Therefore when I put the GitHub syntax for Mermaid it failed. This is a Quatro project so using `.qmd` and the `GitHub markdown` syntax would resolve adding mermaid diagrams.

**Reply by cderv** [reply]
Quarto Markdown syntax is based on Pandoc's Markdown syntax + some addition. We says in https://quarto.org/docs/authoring/markdown-basics.html
> Quarto is based on Pandoc and uses its variation of markdown as its underlying document syntax. Pandoc markdown is an extended and slightly revised version of John Gruber’s [Markdown](https://daringfireball.net/projects/markdown/) syntax.

And we point to [Pandoc's Markdown reference](https://pandoc.org/MANUAL.html#pandocs-markdown). 

This is a different syntax than Github Flavored Markdown, and GFM is an output that you can produce using a `.qmd` file https://quarto.org/docs/output-formats/gfm.html . This means you can write Pandoc's Markdown, and get a compatible Github Document. 


Regarding the mermaid feature, Github has introduced the feature in the HTML previewer to know what to do with mermaid code chunks. They use the source code chunk syntax. 
In quarto, computation code are with computation engine, and we provide a `mermaid` computation that we'll know what to do according to the output chosen. (e.g. for HTML it will create the HTML code and also add in the document all the CSS  and JS required). 

If you want `quarto render` to know what to do with your mermaid chunks, it needs to follow Quarto expected syntax. As you start from a GFM document, you need somehow to convert some syntax. I think you should not have much to convert and this mermaid thing could be the only thing. 
A simple approach could be regex preprocessing to replace `` ```mermaid `` by `` ```{mermaid} `` ), but if you find you need to convert more syntax then you could use pandoc directly to convert from gfm to Pandoc's markdown. 

Hopefully this clarifies better the context and give you ideas.

**Reply by ghost** [reply]
Thank you both for taking the time to discuss this topic.
